wire-ios-data-model/Source/Utilis/EAR/EARService.swift

```diff
@@ -0,0 +1,410 @@
+            let id = secondaryPrivateKeyDescription.id
+            let keyPair = try keyGenerator.generatePublicPrivateKeyPair(id: id)
+            secondaryPublicKey = keyPair.publicKey
+        } catch {
+            WireLogger.ear.error("failed to generate secondary public private keypair: \(String(describing: error))")
+            throw error
+        }
+
+        do {
+            databaseKey = try keyGenerator.generateKey(numberOfBytes: 32)
+            encryptedDatabaseKey = try keyEncryptor.encryptDatabaseKey(
+                databaseKey,
+                publicKey: primaryPublicKey
+            )
+        } catch {
+            WireLogger.ear.error("failed to generate database key: \(String(describing: error))")
+            throw error
+        }
+
+        do {
+            try keyRepository.storePublicKey(
+                description: primaryPublicKeyDescription,
+                key: primaryPublicKey
+            )
+
+            try keyRepository.storePublicKey(
+                description: secondaryPublicKeyDescription,
+                key: secondaryPublicKey
+            )
+
+            try keyRepository.storeDatabaseKey(
+                description: databaseKeyDescription,
+                key: encryptedDatabaseKey
+            )
+        } catch {
+            WireLogger.ear.error("failed to store keys: \(String(describing: error))")
+            throw error
+        }
+
+        return VolatileData(from: databaseKey)
+    }
+
+    // MARK: - Public keys
+
+    public func fetchPublicKeys() throws -> EARPublicKeys {
+        do {
+            return EARPublicKeys(
+                primary: try fetchPrimaryPublicKey(),
+                secondary: try fetchSecondaryPublicKey()
+            )
+        } catch {
+            WireLogger.ear.error("unable to fetch public keys: \(String(describing: error))")
+            throw error
+        }
+    }
+
+    private func fetchPrimaryPublicKey() throws -> SecKey {
+        return try keyRepository.fetchPublicKey(description: primaryPublicKeyDescription)
+    }
+
+    private func fetchSecondaryPublicKey() throws -> SecKey {
+        return try keyRepository.fetchPublicKey(description: secondaryPublicKeyDescription)
+    }
+
+    // MARK: - Private keys
+
+    public func fetchPrivateKeys() throws -> EARPrivateKeys {
+        do {
+            return EARPrivateKeys(
+                primary: try? fetchPrimaryPrivateKey(),
+                secondary: try fetchSecondaryPrivateKey()
+            )
+        } catch {
+            WireLogger.ear.error("unable to fetch private keys: \(String(describing: error))")
+            throw error
+        }
+    }
+
+    private func fetchPrimaryPrivateKey() throws -> SecKey {
+        return try keyRepository.fetchPrivateKey(description: primaryPrivateKeyDescription)
+    }
+
+    private func fetchSecondaryPrivateKey() throws -> SecKey {
+        return try keyRepository.fetchPrivateKey(description: secondaryPrivateKeyDescription)
+    }
```

Comment:
when are the private keys stored ?

Reply:
They are stored when they are created in the Secure Enclave.